marconi/test/Helpers.hs

```diff
@@ -0,0 +1,243 @@
+import Data.Set qualified as Set
+import GHC.Stack qualified as GHC
+import Streaming.Prelude qualified as S
+import System.Directory qualified as IO
+import System.Environment qualified as IO
+import System.FilePath ((</>))
+import System.IO.Temp qualified as IO
+import System.Info qualified as IO
+
+import Hedgehog (MonadTest)
+import Hedgehog qualified as H
+import Hedgehog.Extras.Stock.IO.Network.Sprocket qualified as IO
+import Hedgehog.Extras.Test qualified as HE
+import Hedgehog.Extras.Test.Base qualified as H
+
+import Cardano.Api qualified as C
+import Cardano.Api.Shelley qualified as C
+import Cardano.Streaming qualified as CS
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type (SubmitResult (SubmitFail, SubmitSuccess))
+import Test.Runtime qualified as TN
+import Testnet.Cardano qualified as TN
+import Testnet.Conf qualified as TC (Conf (..), ProjectBase (ProjectBase), YamlFilePath (YamlFilePath), mkConf)
+
+-- * Start testnet
+
+startTestnet
+  :: TN.TestnetOptions
+  -> FilePath
+  -> FilePath
+  -> H.Integration (C.LocalNodeConnectInfo C.CardanoMode, TC.Conf, TN.TestnetRuntime)
+startTestnet testnetOptions base tempAbsBasePath' = do
+  configurationTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml"
+  conf :: TC.Conf <- HE.noteShowM $
+    TC.mkConf (TC.ProjectBase base) (TC.YamlFilePath configurationTemplate)
+      (tempAbsBasePath' <> "/")
+      Nothing
+  tn <- TN.testnet testnetOptions conf
+
+  -- Boilerplate codecs used for protocol serialisation.  The number
+  -- of epochSlots is specific to each blockchain instance. This value
+  -- what the cardano main and testnet uses. Only applies to the Byron
+  -- era.
+  socketPathAbs <- getSocketPathAbs conf tn
+  let epochSlots = C.EpochSlots 21600
+      localNodeConnectInfo =
+        C.LocalNodeConnectInfo
+          { C.localConsensusModeParams = C.CardanoModeParams epochSlots
+          , C.localNodeNetworkId = getNetworkId tn
+          , C.localNodeSocketPath = socketPathAbs
+          }
+
+  pure (localNodeConnectInfo, conf, tn)
+
+getNetworkId :: TN.TestnetRuntime -> C.NetworkId
+getNetworkId tn = C.Testnet $ C.NetworkMagic $ fromIntegral (TN.testnetMagic tn)
+
+getSocketPathAbs :: (MonadTest m, MonadIO m) => TC.Conf -> TN.TestnetRuntime -> m FilePath
+getSocketPathAbs conf tn = do
+  let tempAbsPath = TC.tempAbsPath conf
+  socketPath <- IO.sprocketArgumentName <$> H.headM (TN.nodeSprocket <$> TN.bftNodes tn)
+  socketPathAbs <- H.note =<< (liftIO $ IO.canonicalizePath $ tempAbsPath </> socketPath)
+  pure socketPathAbs
+
+getPoolSocketPathAbs :: (MonadTest m, MonadIO m) => TC.Conf -> TN.TestnetRuntime -> m FilePath
+getPoolSocketPathAbs conf tn = do
+  let tempAbsPath = TC.tempAbsPath conf
+  socketPath <- IO.sprocketArgumentName <$> H.headM (TN.poolNodeSprocket <$> TN.poolNodes tn)
+  socketPathAbs <- H.note =<< (liftIO $ IO.canonicalizePath $ tempAbsPath </> socketPath)
+  pure socketPathAbs
+
+--
```

Comment:
is this an end of section?

Reply:
It was, removed it 84d56ff9fe0f34a95d8f2cb199d99ed1d5cd1008